### PR TITLE
Cumulus 123 mhs

### PR DESCRIPTION
**Pull request by flamingbear**
Dashboard filters and searches can have their state in the URL so that they are shareable.

### CODE REVIEW

File: app/scripts/utils/url-helper.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  if (value.length) {
+    nextQuery = { ...nextQuery, [paramKey]: value };
+  } else {
+    delete nextQuery[paramKey];
+  }
+  if (location.query[paramKey] !== nextQuery[paramKey]) {
+    location.query = nextQuery;
+    router.push(location);
+  }
+}
```
Comment (flamingbear) [inline app/scripts/utils/url-helper.js]: These could probably use some tests....

File: app/scripts/components/table/list-view.js [hunk 2]
```diff
@@ -144,7 +143,7 @@ class List extends React.Component {
       page: this.state.page,
       order: this.state.order,
       sort_by: this.getSortProp(this.state.sortIdx),
-      params: this.state.params,
+      ...this.state.params,
```
Comment (markdboyd) [inline app/scripts/components/table/list-view.js]: Are these equivalent? The former operation would place the contents of `this.state.params` on the `params` property, whereas the spread operation will put the contents of `this.state.params` directly on the object.

File: app/scripts/utils/status.js [hunk 3]
```diff
@@ -6,22 +6,9 @@ const status = {
 };
 export default status;
 
-export const queryStatus = [
-  'running',
-  'completed',
-  'failed'
-];
-
```
Comment (markdboyd) [inline app/scripts/utils/status.js]: Why is this getting deleted? Unused?

File: cypress/integration/granules_spec.js [hunk 4]
```diff
@@ -107,16 +109,52 @@ describe('Dashboard Granules Page', () => {
...
+      cy.get('#form-Status-status > div > input').as('status-input');
+      cy.get('@status-input').should('have.value', 'Completed');
+    });
+
+    it('Should update dropdown with status value if query params are not a valid status.', () => {
+      cy.visit('/#/granules?status=notanoption');
+      cy.get('#form-Status-status > div > input').as('status-input');
+      cy.get('@status-input').should('have.value', 'notanoption');
+    });
+
```
Comment (markdboyd) [inline cypress/integration/granules_spec.js]: Hmm, is this a behavior that we want? That entering arbitrary values in the URL actually show up in the component?

File: cypress/integration/granules_spec.js [hunk 5]
```diff
@@ -107,16 +109,52 @@ describe('Dashboard Granules Page', () => {
...
+    it('Should update URL when dropdown filters are activated.', () => {
+      visitGranulePage();
+      cy.get('#form-Status-status > div > input').as('status-input');
+      cy.get('@status-input').click().type('{downarrow}').type('{enter}');
+      cy.url().should('include', '?status=running');
```
Comment (markdboyd) [inline cypress/integration/granules_spec.js]: I don't love that this test assumes an order, but don't see any alternative

File: cypress/integration/granules_spec.js [hunk 6]
```diff
@@ -107,16 +109,52 @@ describe('Dashboard Granules Page', () => {
...
+    });
+
+    it('Should show Search and Dropdown filters in URL.', () => {
+      visitGranulePage();
+      cy.get('.search').as('search');
+      cy.get('@search').click().type('L2');
+      cy.get('#form-Status-status > div > input').as('status-input');
+      cy.get('@status-input').click().type('{downarrow}{downarrow}{enter}');
+      cy.url().should('include', 'search=L2');
+      cy.url().should('include', 'status=completed');
```
Comment (markdboyd) [inline cypress/integration/granules_spec.js]: I think you could write this as

`cy.url().should('include', 'search=L2').and('include', 'status=completed')`

File: .eslintrc [hunk 7]
```diff
@@ -6,7 +6,8 @@
       "mocha": true
     },
     "plugins": [
-      "react"
+      "react",
+      "react-hooks"
```
Comment (markdboyd) [inline .eslintrc]: Are we using React hooks anywhere in the code now?

File: app/scripts/components/form/dropdown.js [hunk 8]
```diff
@@ -21,39 +23,68 @@ function renderMenu (items, value, style) {
...
-      value: ''
+      value: initialValue
     };
     this.onSelect = this.onSelect.bind(this);
     this.onChange = this.onChange.bind(this);
   }
 
+  componentDidUpdate (prevProps, prevState, snapshot) {
+    const {location, paramKey, dispatch, action, options} = this.props;
+    if (location.query[paramKey] !== prevProps.location.query[paramKey]) {
```
Comment (markdboyd) [inline app/scripts/components/form/dropdown.js]: Hmm, is `location.query` always defined, even on initial component load? I'm surprised that this never errors.

File: app/scripts/components/form/dropdown.js [hunk 9]
```diff
@@ -21,39 +23,68 @@ function renderMenu (items, value, style) {
...
+      let value = location.query[paramKey];
+      dispatch(action({ key: paramKey, value }));
+      if (value) value = statusToLabel(options, value);
+      this.setState({value}); // eslint-disable-line react/no-did-update-set-state
```
Comment (markdboyd) [inline app/scripts/components/form/dropdown.js]: You only want to do `this.setState` if you got a value from `statusToLabel`, correct? So this should this statement be inside the `if` block?

File: app/scripts/utils/url-helper.js (hunk 1 above)
Comment (markdboyd) [reply app/scripts/utils/url-helper.js]: Yeah, I wonder if there's a package providing middleware that would handle this for us. But we can discuss this offline.

File: test/reducers/test-removeDeleted.js [hunk 10]
```diff
@@ -0,0 +1,41 @@
+'use strict';
```
Comment (markdboyd) [inline test/reducers/test-removeDeleted.js]: Very nice. Love to see additional tests 👍

File: app/scripts/components/form/search.js [hunk 11]
```diff
@@ -48,20 +64,27 @@ class Search extends React.Component {
 
   render () {
     return (
-      <div className='filter__item'>
-        <form className='search__wrapper form-group__element' onSubmit={this.submit} value={this.state.value}>
-          <input className='search' type='search' onChange={this.complete}/>
-          <span className='search__icon'/>
+        <div className='filter__item'>
```
Comment (markdboyd) [inline app/scripts/components/form/search.js]: Why remove the indentation on this line and line 72?

File: .eslintrc (hunk 7 above)
Comment (flamingbear) [reply .eslintrc]: I updated to a version that will take hooks, and included the hook linter to be prepared.

File: app/scripts/utils/url-helper.js (hunk 1 above)
Comment (flamingbear) [reply app/scripts/utils/url-helper.js]: def willing to entertain different ideas.

File: app/scripts/components/table/list-view.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (flamingbear) [reply app/scripts/components/table/list-view.js]: No, they're not equivalent and the added spread fixes a bug where filters weren't working.

File: app/scripts/utils/status.js (hunk 3 above)
Comment (flamingbear) [reply app/scripts/utils/status.js]: I couldn't find any use of those variables in our code, so I just deleted because they were confusing me.  If you can think of a reason to keep them, please let me know.

File: cypress/integration/granules_spec.js (hunk 4 above)
Comment (flamingbear) [reply cypress/integration/granules_spec.js]: It would be the equivalent if you typed something in the dropdown box that didn't match any of the configured options.  I don't know if we want to prevent that or not?

Ok, I just actually tested that. and if you type a `z` in the dropdown and it doesn't match anything, nothing is kicked off, and if you hit return (select), the url pushes a ? before the hash so we get `http://localhost:3000/?#/granules` . That doesn't seem right.  

I'll have to figure out how to handle this with search params differently from dropdowns.  But Ok I can see this might not be what we want.

To clarify:  you think that if you browse to a url that has a invalid status key, the dropdown corresponding to that paramKey should clear itself, and remove that from the query?

so basically L133 should be `      cy.get('@status-input').should.not('have.value', 'notanoption');`

File: cypress/integration/granules_spec.js (hunk 5 above)
Comment (flamingbear) [reply cypress/integration/granules_spec.js]: Let me try something.

File: app/scripts/components/form/dropdown.js (hunk 8 above)
Comment (flamingbear) [reply app/scripts/components/form/dropdown.js]: By creating the router history with `useRouterHistory(createHashHistory)()`, there will always be a query object on a location.   It may be empty, but then these are undefined and not errors.

File: app/scripts/utils/status.js (hunk 3 above)
Comment (markdboyd) [reply app/scripts/utils/status.js]: Nope, that's fine with me

File: app/scripts/components/form/dropdown.js (hunk 9 above)
Comment (flamingbear) [reply app/scripts/components/form/dropdown.js]: No, because we always want to set the state's value no matter what the user has typed, but if the user types running, we will correct it to Running.

And if the user deleted everything in the box, we want to set the value to the empty string.

File: cypress/integration/granules_spec.js (hunk 4 above)
Comment (markdboyd) [reply cypress/integration/granules_spec.js]: I think if the URL has an invalid status key, basically nothing should happen, so whatever logic initially sets the value should short-circuit if the key is invalid.

We probably also don't want the case you described where typing a invalid value in a dropdown and hitting enter affects the URL

File: app/scripts/components/form/search.js (hunk 11 above)
Comment (flamingbear) [reply app/scripts/components/form/search.js]: emacs. autoformating problems I would guess.

File: app/scripts/components/form/dropdown.js (hunk 9 above)
Comment (markdboyd) [reply app/scripts/components/form/dropdown.js]: Ah, got it

File: test/reducers/test-removeDeleted.js (hunk 10 above)
Comment (flamingbear) [reply test/reducers/test-removeDeleted.js]: Well, I had an error in that and I had no idea why, so had to figure out what it was doing.

File: cypress/integration/granules_spec.js (hunk 4 above)
Comment (flamingbear) [reply cypress/integration/granules_spec.js]: What should the state of the URL be after discovering a bad value?
say `?status=badvalue`, should that be removed to match the dropdown value that is also blank?

In the case of the problem with the URL updating badly, that's an existing bug that I filed a ticket for https://bugs.earthdata.nasa.gov/browse/CUMULUS-1621, but I'm hoping that it will go away in an update to react router or an update to a browser history instead of a hash history.

File: app/scripts/components/form/search.js (hunk 11 above)
Comment (flamingbear) [reply app/scripts/components/form/search.js]: e382c25

File: cypress/integration/granules_spec.js (hunk 5 above)
Comment (flamingbear) [reply cypress/integration/granules_spec.js]: 5e5ef70

File: cypress/integration/granules_spec.js (hunk 6 above)
Comment (flamingbear) [reply cypress/integration/granules_spec.js]: f16adad
